RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerGlobal.cc

```diff
@@ -18,7 +18,7 @@
 #include <FWCore/ParameterSet/interface/ParameterSetDescription.h>
 
 EcalUncalibRecHitWorkerGlobal::EcalUncalibRecHitWorkerGlobal(const edm::ParameterSet&ps,edm::ConsumesCollector& c) :
-  EcalUncalibRecHitWorkerRunOneDigiBase(ps,c), useDBShape(false), testbeamEEShape(EEShape(useDBShape)), testbeamEBShape(EBShape(useDBShape))
+  EcalUncalibRecHitWorkerRunOneDigiBase(ps,c), useDBShape(true), testbeamEEShape(EEShape(useDBShape)), testbeamEBShape(EBShape(useDBShape))
```

Comment:
this and other changes are hardcoded without an easy option to change back.

I just wanted to make sure that there will be no more cases from ECAL DPG where the shape will be needed in the old way.
If that's the case, can we already plan to eliminate the branches of the code which still check useDBShape and pretend to be able to do something with this value set to ```false```?

Reply:
@slava77 
There are two parts of the software using false deliberately, for the shake of simplicity. 

a) the testbeam software dating back to 2004-2007 which is used to reconstruct/analyze the testbeam data. For those old data, I can't think of any reason changing the pulse shape, since it was these data that were actually used to derive the pulse shapes we used since then.  Having useShapeDB = false, there it's OK in my opinion and keeps things simple.

b) a unit test that hasn't access to the EventSetup and couldn't figure out how to pass the pointer to it. However, this unit test is supposed to test the shape-building-logic and not the input shape per se, which could be a dummy one, testing the consistency between input/output. In this part, I also think that not using the DB conditions is just making simpler the software since the purpose of the test -- which is to catch failures -- in the shape building logic is achieved even if a dummy set of shapes is used.

the rest of the CMSSW, needs to useShapeDB = true

Reply:
> a) the testbeam software dating back to 2004-2007 which is used to reconstruct/analyze the testbeam data

IMHO, this update to using DB is about code modernization and not about preserving semantics of the code written in 2004.
IIUC, the testbeam part can also be loaded from DB.

Is it practical to remove ```useDBShape``` now, at least in the reco code (and drop the conditionals and side branches of the code where ```false``` is still an option)?
From your description it sounds like there is no need for this functionality anymore.

Reply:
@slava77 
trying to think of myself as somebody who would want to analyze testbeam data, I would just find annoying to figure out that somebody in 2018 disabled a functionality that was always been there and added an external dependency to DB to do the same work.

A technical issue, removing the useDBShape functionality will also result to a break of a unit test

https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/EcalSimAlgos/test/testEcalShape.cpp

which I don't know how to modernize and make it also dependent in the EventSetup. If the suppression of the useDBShape flag is a strong condition in order to move forward, I will need code examples of how the unit-test can be updated to have a pointer to the EventSetup.

The addition of this Boolean flag was suggested to me by DPG people, initially I was very against of having 2-branches logic. In the end, I coded as it was suggested to me because of the testbeam package simplicity and the unit test breakage. An additional good aspect IMHO, the branches describing what happens in the false option stand also as form documentation of what were the legacy shapes used in phase-I. Sure, those will from now on be loaded in DB, but it's not the same easiness for the average users of this software to find out what is in a DB compared to just read from their browser the coded shapes in the repository.

Reply:
perhaps I'm missing where the TB analysis  code is located.
If it is used in the Reco\* packages modified in this PR, then either this useDBShape in these packages has to be made a configurable or it should be eliminated together with adding TB shape to the DB for use in the Reco\* packages.

Note though that I do not have a particularly strong opinion about specifics of the code in /test/ directories. FWIW, in most cases it should just compile.

Reply:
> perhaps I'm missing where the TB analysis code is located.

https://github.com/cms-sw/cmssw/blob/5333441168f3588b8d205f4bb9459f7e6ed6e119/RecoTBCalo/EcalTBRecProducers/src/EcalTBWeightUncalibRecHitProducer.cc#L49


> 
> Note though that I do not have a particularly strong opinion about specifics of the code in /test/ directories. FWIW, in most cases it should just compile.

it compiles, but it crashes on execution if a valid EventSetup pointer is not available within main() to be passed to the EcalShape class.  see 
https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/EcalSimAlgos/test/testEcalShape.cpp

Reply:
OK, I see that RecoTBCalo is not touched in this PR.
Does this mean that all other Reco\* packages modified in this PR will not need the TB data inputs/shapes?
If so, I think that the ```useDBShape``` can be eliminated in these files

Reply:
@slava77 

I am not sure I understand the question. How exactly you suggest to eliminate the useDBShape ? If it is at level of the constructor, i.e., making it to assume always "true" and erase the "false" from being an option, then the TB package and the unitest will break unless they are also updated. If those are to be updated, let me know how the unit test can be made such as to have a pointer to the EventSetup https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/EcalSimAlgos/test/testEcalShape.cpp 

In my view, the best is to leave the useDBShape functionality for reasons of simplicity and documentation of how the software was working in phase-I. 

What problems do you foresee by this extra degree of freedom that the present software has, i.e., ability to decide to use (or not) the DB) ?

Reply:
@slava77 
is your suggestion the following one ?

Erase  
https://github.com/theofil/cmssw/blob/07aca385521efaa41d90eefee44706c5e531ceb1/RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerGlobal.h#L69

 and then update 

RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerGlobal.cc

https://github.com/theofil/cmssw/blob/07aca385521efaa41d90eefee44706c5e531ceb1/RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitWorkerGlobal.cc#L21

such as  the
<img width="1078" alt="image" src="https://user-images.githubusercontent.com/4677362/45480669-cfb45e00-b749-11e8-9333-e61932bcb3fa.png">

becomes 

EcalUncalibRecHitWorkerRunOneDigiBase(ps,c), testbeamEEShape(EEShape(true)), testbeamEBShape(EBShape(true))

and similar for the L65  without touching EBShape, EEShape, APDShape and EcalShapeBase classes ?

IF this is your suggestion, then I think it is straightforward to implement it.

Reply:
yes.
Sorry if I was not clear in my earlier comments.

Reply:
@slava77 

I have initially understood that you was suggesting to eliminate the useDBShape from the EcalShape class and thus the eliminate the possibility to instantiate the legacy phase-I shapes without using DB conditions, this would have caused a sequence of problems.

I updated the PR to reflect better your suggestion, please have a look.